### ISSUE TITLE
Bad LOOP macro segfaults

### ISSUE THREAD

**Issue by drmeister** [clasp-developers/clasp#1239]
```
(defun screw-up ()
  (loop for x in 42
        do (format t "x = ~a~%" x)))
```
Will segfault.  Nothing should segfault.

**Comment by kpoeck** [comment]
Might be the the same as #1212.

**Comment by kpoeck** [comment]
I believe the issue is that loop expands into code using CORE:CONS-CAR which assumes that the parameter is a cons, which it isn't in this case.
```lisp
(macroexpand ' (loop for x in 42
        do (format t "x = ~a~%" x)))

(BLOCK NIL
  (LET ((X NIL) (#:LOOP-LIST2651 42))
    (DECLARE (TYPE (OR LIST (INTEGER 42 42)) #:LOOP-LIST2651)
     (IGNORABLE #:LOOP-LIST2651) (IGNORABLE X))
    (CORE::LOOP-BODY NIL
                     ((WHEN (ENDP #:LOOP-LIST2651) (GO CORE::END-LOOP))
                      (CORE::LOOP-REALLY-DESETQ X
                                                (CORE:CONS-CAR
                                                 #:LOOP-LIST2651))
                      NIL
                      (CORE::LOOP-REALLY-DESETQ #:LOOP-LIST2651
                                                (CORE:CONS-CDR
                                                 #:LOOP-LIST2651)))
                     ((FORMAT T "x = ~a~%" X))
                     ((WHEN (ENDP #:LOOP-LIST2651) (GO CORE::END-LOOP))
                      (CORE::LOOP-REALLY-DESETQ X
                                                (CORE:CONS-CAR
                                                 #:LOOP-LIST2651))
                      NIL
                      (CORE::LOOP-REALLY-DESETQ #:LOOP-LIST2651
                                                (CORE:CONS-CDR
                                                 #:LOOP-LIST2651)))
                     NIL)))
````
with core:car-cons being
```lisp
(fset 'cons-car #'(lambda (def env)
                    (declare (ignore env))
                    `(car (the cons ,(cadr def))))
      t)
````

**Comment by kpoeck** [comment]
The right fix is not to use core:cons-car in loop, unless it is proven that the argument is a cons.

I have this lying around, so let me do a PR for loop

**Comment by Bike** [comment]
I'm working on some compiler changes right now that should make declarations less unsafe. Since even if we change loop, we probably shouldn't segfault on `(car (the cons x))` unless *maybe* safety is zero. With the changes, the declaration will still be useful (e.g. code paths premised on the variable being null could be deleted), but clasp won't believe it naively.

**Comment by Bike** [comment]
those changes are now merged in b0995bcdcb6dbabcb82e1626000a1d0f83d2d191

**Comment by kpoeck** [comment]
gives now a nice error message so certainly does not segfault.

i'd still pitch for my pr 1241, altough it suffered collateral damage from the lsp -> lisp renaming
